src/RevCPU.cc

```diff
@@ -2651,95 +2654,102 @@ void RevCPU::UpdateThreadAssignments(uint32_t ProcID){
+      AssignedThreads.at(ProcID).erase(Thread->GetThreadID());
+      CompletedThreads.emplace(Thread->GetThreadID());
+      if( AssignedThreads.at(ProcID).empty() ){
+        Enabled[ProcID] = false;
       }
-      else {
-        // This thread didn't experience a state change
-        continue;
+      break;
+    case ThreadState::BLOCKED:
+      // This thread is blocked (currently only caused by a rev_pthread_join)
+      // We need to:
+      // 1. Check if the thread it is waiting on has already completed
+      // 2. If it has... Thread can resume execution
+      // 3. If not, thread remains blocked
+      //    3a. Move its ThreadID to the BlockedThreads list
+      //    3b. Remove it from the AssignedThreads lis
+      output.verbose(CALL_INFO, 8, 0, "Thread %" PRIu32 "on Core %" PRIu32 " is BLOCKED\n", Thread->GetThreadID(), ProcID);
+      // -- 1.
+      if( ThreadCanProceed(Thread->GetThreadID()) ){
+        // -- 2.
+        output.verbose(CALL_INFO, 8, 0, "Thread %" PRIu32 " on Core %" PRIu32 " was waiting on thread %u which has already completed so it can proceed\n",
+                        Thread->GetThreadID(), ProcID, Thread->GetWaitingToJoinTID());
+        // Continue executing thread on same Core
+        Thread->SetState(ThreadState::RUNNING);
       }
+      else { // -- 3.
+        output.verbose(CALL_INFO, 8, 0, "Thread %" PRIu32 " on Core %" PRIu32 " was waiting on thread %u which has not yet completed so it remains blocked\n",
+                        Thread->GetThreadID(), ProcID, Thread->GetWaitingToJoinTID());
+        Thread->SetState(ThreadState::BLOCKED);
+        // -- 3a.
+        BlockedThreads.emplace(Thread->GetThreadID());
+
+        std::cout << Thread->to_string();
+        
+        // -- 3b.
+        AssignedThreads.at(ProcID).erase(Thread->GetThreadID());
+        
+        if( AssignedThreads.at(ProcID).empty() ){
+          Enabled[ProcID] = false;
+        }
+
+      }
+      break;
+    case ThreadState::START: // Should never happen
+      output.verbose(CALL_INFO, 99, 1, "A new thread with ID = %" PRIu32 " was found on Core %" PRIu16, Thread->GetThreadID(), ProcID);
+
+      // Mark it ready for execution
+      Thread->SetState(ThreadState::READY);
+
+      // Add it to the Thread map
+      Threads.emplace(Thread->GetThreadID(), Thread);
+
+      // Add it to the thread queue to be scheduled
+      ThreadQueue.emplace_back(Thread->GetThreadID());
+      break;
+
+    case ThreadState::RUNNING:
+      output.verbose(CALL_INFO, 11, 0, "Thread %" PRIu32 " on Core %" PRIu32 " is RUNNING\n", Thread->GetThreadID(), ProcID);
+      break;
+
+    case ThreadState::READY:
+      // If this happens we are not setting state when assigning thread somewhere
+      output.fatal(CALL_INFO, 99, "Error: Thread %" PRIu32 " on Core %" PRIu32 " is assigned but is in START state... This is a bug\n",
```

Comment:
Should this read `START state..` or `READY state...`

Reply:
Sure should!